source/validate.h

```diff
@@ -43,6 +45,26 @@
 #include "spirv_definition.h"
 #include "table.h"
 
+// clang-format off
+#define _printf_  printf
+#define MSG(msg,...) do {                       \
+    _printf_(__FILE__":%d: " msg "\n",          \
+             __LINE__, ##__VA_ARGS__);          \
+  } while (0)
+
+#ifdef DEBUG
+#define MSG_DBG      MSG
+#else
+#define MSG_DBG(...) do { /* nothing */ } while (0)
+#endif
+
+
+#define SHOW(exp)   do { MSG("%s  %ld",   #exp, (long)exp);   } while (0)
+#define SHOW_P(ptr) do { MSG("%s  %#zx",  #ptr, (size_t)ptr); } while (0)
+#define SHOW_F(exp) do { MSG("%s  %g",    #exp, (double)exp); } while (0)
+#define SHOW_S(exp) do { MSG("%s  %s",    #exp, exp);         } while (0)
+// clang-format on
```

Comment:
So keep the top-level macros to capture `__FILE__` and `__LINE__`, but use functions below them.  I feel that the `MSG` macros are difficult to understand. Will the variadic versions ever be invoked with no arguments (which the standard prohibits)? Will the `MSG_DBG` replacement rescan correctly to become a function-like invocation of `MSG`? I'm sure there's more questions there, and I wouldn't want to debug through it on an unfamiliar compiler.  Not to mention that macros don't typecheck their arguments and pollute scopes surrounding their invocations.